### PR TITLE
Novitsky/backup for elasticsearch

### PR DESCRIPTION
**Pull request by SergeyNovitsky**
First example of implementation

### CODE REVIEW

File: OutOfSchool/OutOfSchool.WebApi/Services/BackupTrackerService.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            this.logger = logger;
+            this.localazer = localizer;
+        }
+
+        public async Task<BackupTrackerDto> Create(BackupTrackerDto dto)
+        {
+            var backupTracker = dto.ToDomain();
+            var newBackupTracker = await repository.Create(backupTracker).ConfigureAwait(false);
+            return newBackupTracker.ToModel();
+        }
```
Comment (yfedo) [inline OutOfSchool/OutOfSchool.WebApi/Services/BackupTrackerService.cs:29]: do you really need this return result newBackupTracker.ToModel()?))
as for now it is not used))

File: OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            foreach (var backupTrackerRecord in result)
+            {
+                workshop = null;
+                esResultIsValid = false;
+                switch (backupTrackerRecord.Operation)
+                {
+                    case BackupOperation.Create:
+                        workshop = await databaseService.GetById(backupTrackerRecord.RecordId).ConfigureAwait(false);
+                        esResultIsValid = await elasticsearchService.Index(workshop.ToESModel()).ConfigureAwait(false);
+                        break;
```
Comment (yfedo) [inline OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs:52]: what will happen in case of failure to index here?))

File: OutOfSchool/OutOfSchool.DataAccess/Repository/WorkshopRepository.cs [hunk 3]
```diff
@@ -45,5 +46,23 @@ public new async Task Delete(Workshop entity)
...
+                                 group record by record.RecordId into groupedRecords
+                                 select new { RecordId = groupedRecords.Key, MaxOperationDate = groupedRecords.Max(r => r.OperationDate) };
+
+            var result = from rMaxDate in resultMaxDates
+                         join record in elasticsearchSyncRecords on new { rMaxDate.RecordId, OperationDate = rMaxDate.MaxOperationDate } equals new { record.RecordId, record.OperationDate } into leftJoin
+                         from joinedRecord in leftJoin
+                         where joinedRecord.Operation == operation
+                         select new { joinedRecord.RecordId, joinedRecord.OperationDate, joinedRecord.Operation };
+
+            return await db.Workshops.Join(result, workshop => workshop.Id, result => result.RecordId, (workshop, result) => workshop).ToListAsync();
```
Comment (VadymLevkovskyi) [inline OutOfSchool/OutOfSchool.DataAccess/Repository/WorkshopRepository.cs:65]: 1. From join I see that `RecordId` is FK field to primary key `WorkshopId`. 
2. I don't see this in migration script.

So, either naming is wrong (and `RecordId` should be named as `WorkshopId` and db schema should be fixed) or extra filtering condition is missed (that has to specify that only workshops' data are taken from `ElasticsearchSyncRecords`)

Similar logic is present in `WorkshopServicesCombiner` - where workshop id is stored as record id without any additional type info

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs [hunk 4]
```diff
@@ -49,11 +57,15 @@ public async Task<WorkshopDTO> Update(WorkshopDTO dto)
         {
             var workshop = await databaseService.Update(dto).ConfigureAwait(false);
 
-            var esResultIsValid = await elasticsearchService.Update(workshop.ToESModel()).ConfigureAwait(false);
+            //var esResultIsValid = await elasticsearchService.Update(workshop.ToESModel()).ConfigureAwait(false);
+
+            var esResultIsValid = false;
 
             if (!esResultIsValid)
             {
```
Comment (VadymLevkovskyi) [inline OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:65]: probably this code should be swept

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs [hunk 5]
```diff
@@ -64,11 +76,15 @@ public async Task Delete(long id)
         {
             await databaseService.Delete(id).ConfigureAwait(false);
 
-            var esResultIsValid = await elasticsearchService.Delete(id).ConfigureAwait(false);
+            //var esResultIsValid = await elasticsearchService.Delete(id).ConfigureAwait(false);
+
+            var esResultIsValid = false;
```
Comment (VadymLevkovskyi) [inline OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:81]: another testing? artifact

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs (hunk 4 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:65]: it is only for test reason. when I will be ready for merge I will change it.

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs (hunk 5 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:81]: yes, the same.

File: OutOfSchool/OutOfSchool.WebApi/Services/BackupTrackerService.cs (hunk 1 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.WebApi/Services/BackupTrackerService.cs]: no, I don't. I will fix it after I will be sure about structure of the method.

File: OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs (hunk 2 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs]: yes, I implemented only ideal situation now, I will fix it.

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs (hunk 4 above)
Comment (VadymLevkovskyi) [reply OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:65]: Pls, write `//TODO` comments for such cases, it will be helpful both for you (just Ctrl-F all TODOs and nothing is missed) and for reviewers to understand you intention

File: OutOfSchool/OutOfSchool.DataAccess/Repository/WorkshopRepository.cs [hunk 6]
```diff
@@ -45,5 +46,23 @@ public new async Task Delete(Workshop entity)
...
+            var elasticsearchSyncRecords = db.ElasticsearchSyncRecords;
+
+            var resultMaxDates = from record in elasticsearchSyncRecords
+                                 group record by record.RecordId into groupedRecords
+                                 select new { RecordId = groupedRecords.Key, MaxOperationDate = groupedRecords.Max(r => r.OperationDate) };
+
+            var result = from rMaxDate in resultMaxDates
+                         join record in elasticsearchSyncRecords on new { rMaxDate.RecordId, OperationDate = rMaxDate.MaxOperationDate } equals new { record.RecordId, record.OperationDate } into leftJoin
+                         from joinedRecord in leftJoin
+                         where joinedRecord.Operation == operation
```
[the remark this answers is not in the document]
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.DataAccess/Repository/WorkshopRepository.cs:62]: at first part of query I take latest records by date and I don't need filter it by operation beforehand because it possible that one I could be created, updated or deleted during this period. then I join with the same table for getting not aggregated information (operation in this case). It was first iteration of implementation of this query. I will think about its optimization.

File: OutOfSchool/OutOfSchool.DataAccess/Repository/WorkshopRepository.cs (hunk 3 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.DataAccess/Repository/WorkshopRepository.cs:65]: yes, at first I added additional field "TableName", I consider that in feature there are possible different entities, but for now I decided to simplify table. not sure that that was a correct decision. I will fix it.

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs (hunk 4 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:65]: thank you. I will.

File: OutOfSchool/OutOfSchool.WebApi/Extensions/MappingExtensions.cs [hunk 7]
```diff
@@ -166,6 +167,11 @@ public static WorkshopDTO ToModelSimple(this Workshop workshop)
             });
         }
 
+        public static ElasticsearchSyncRecordDto ToModel(this ElasticsearchSyncRecord elasticsearchSyncRecord)
```
Comment (VadymLevkovskyi) [inline OutOfSchool/OutOfSchool.WebApi/Extensions/MappingExtensions.cs:170]: Approach with constructing Mapper on-the-fly is deprecated (due to perf. reason). Please inject mapper dependency and configure AutoMapper on Startup

File: OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        {
+            this.databaseService = workshopService;
+            this.repository = repository;
+            this.esProvider = esProvider;
+            this.logger = logger;
+        }
+
+        public async Task<bool> Synchronize()
+        {
+            {
```
Comment (VadymLevkovskyi) [inline OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs:42]: Extra `{}` seems irrelevant

File: OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+                Result result;
```
Comment (VadymLevkovskyi) [inline OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs:43]: Please move declaration of variable closer to it's 1st usage (line 56). 
For variable declaration wherever possible use `var` - it enforces usage of beforementioned rule (in addition to other benefits like explicitness in type conversion and naming).

**Review by DmyMi — APPROVED** [review]

File: OutOfSchool/OutOfSchool.WebApi/Services/Database/WorkshopService.cs [hunk 10]
```diff
@@ -222,6 +222,27 @@ public async Task<SearchResult<WorkshopDTO>> GetByFilter(WorkshopFilter filter =
             return result;
         }
 
+        public async Task<IEnumerable<WorkshopDTO>> GetWorkshopsForCreate()
+        {
+            var workshops = await workshopRepository.GetListOfWorkshopsForSynchronizationByOperation(ElasticsearchSyncOperation.Create).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.WebApi/Services/Database/WorkshopService.cs]: fixed

File: OutOfSchool/OutOfSchool.WebApi/Extensions/MappingExtensions.cs (hunk 7 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.WebApi/Extensions/MappingExtensions.cs]: fixed

File: OutOfSchool/OutOfSchool.WebApi.Tests/Services/WorkshopServicesCombinerTests.cs [hunk 11]
```diff
@@ -43,7 +43,7 @@ public void SetUp()
             mockElasticsearchService = new Mock<IElasticsearchService<WorkshopES, WorkshopFilterES>>();
             mockLogger = new Mock<ILogger<WorkshopServicesCombiner>>();
 
-            service = new WorkshopServicesCombiner(mockDatabaseService.Object, mockElasticsearchService.Object, mockLogger.Object);
+            service = new WorkshopServicesCombiner(mockDatabaseService.Object, mockElasticsearchService.Object, mockLogger.Object, mockElasticsearchSynchronizationService.Object);
```
[the remark this answers is not in the document]
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.WebApi.Tests/Services/WorkshopServicesCombinerTests.cs]: fixed

File: OutOfSchool/OutOfSchool.ElasticsearchData/ElasticsearchProvider.cs [hunk 12]
```diff
@@ -51,6 +52,18 @@ public virtual async Task<Result> DeleteEntityAsync(TEntity entity)
...
         }
 
+        public virtual async Task<Result> DeleteRangeOfEntitiesByIdsAsync(IEnumerable<long> ids)
+        {
+            var bulkResponse = await ElasticClient.BulkAsync(new BulkRequest
+            {
+                Operations = ids.Select(x => new BulkDeleteOperation<TEntity>(x)).Cast<IBulkOperation>().ToList(),
+            });
+
+            Result result = Result.Deleted;
```
[the remark this answers is not in the document]
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.ElasticsearchData/ElasticsearchProvider.cs]: fixed

File: OutOfSchool/OutOfSchool.DataAccess/Repository/EntityRepository.cs [hunk 13]
```diff
@@ -64,6 +64,19 @@ public async Task Delete(T entity)
             await this.dbContext.SaveChangesAsync();
         }
 
+        /// <inheritdoc/>
+        public async Task DeleteAll()
+        {
+            var records = dbSet.Select(x => x);
```
[the remark this answers is not in the document]
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.DataAccess/Repository/EntityRepository.cs]: fixed

File: OutOfSchool/OutOfSchool.WebApi/Controllers/V1/ElasticsearchSynchronizationController.cs [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace OutOfSchool.WebApi.Controllers.V1
+{
+    /// <summary>
+    /// Controller for data synchronization between databases.
+    /// </summary>
+    [ApiController]
+    [ApiVersion("1.0")]
+    [Route("api/v{version:apiVersion}/[controller]/[action]")]
+    [Route("[controller]/[action]")]
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/V1/ElasticsearchSynchronizationController.cs:15]: don't need controller route

File: OutOfSchool/OutOfSchool.WebApi/Controllers/V1/ElasticsearchSynchronizationController.cs [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        /// <response code="400">If the model is invalid, some properties are not set etc.</response>
+        /// <response code="401">If the user is not authorized.</response>
+        /// <response code="403">If the user has no rights to use this method, or sets some properties that are forbidden.</response>
+        /// <response code="500">If any server error occures.</response>
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [HttpGet]
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/V1/ElasticsearchSynchronizationController.cs:41]: probably should change to `POST` request (it's ok to have an empty body), as you are changing the system's state by doing this call. Also using `POST` eliminates a problem with `GET` request's response being cached somewhere in between client and server which might lead to situation where admin wants to sync, gets an `OK` response but the request did not actually reach the server but hit a cache.

File: OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+                logger.LogError($"Syncronization of Elasticsearch has failed during {ElasticsearchSyncOperation.Delete} operation");
+                return false;
+            }
+
+            logger.LogInformation($"Synchronization of Elasticsearch has finished successfully.");
+
+            return true;
+        }
+
+        public async Task Create(ElasticsearchSyncRecordDto dto)
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs:82]: do we really need a DTO? we're using this method in the same class but we spend processor time converting dto-> entity as well as instantiate 2 classes (dto and entity).

File: OutOfSchool/OutOfSchool.WebApi/Services/Database/IWorkshopService.cs [hunk 17]
```diff
@@ -41,5 +41,7 @@ public interface IWorkshopService : ICRUDService<WorkshopDTO, Guid>
         /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.
         /// The task result contains the <see cref="SearchResult{WorkshopCard}"/> that contains found elements.</returns>
         Task<SearchResult<WorkshopCard>> GetNearestByFilter(WorkshopFilter filter = null);
+
+        Task<IEnumerable<WorkshopDTO>> GetByIds(IEnumerable<Guid> ids);
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.WebApi/Services/Database/IWorkshopService.cs:45]: do we really need to return DTO? see `ElasticsearchSynchronizationService.cs` comment

File: OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+
+                if (result == Result.Error)
+                {
+                    logger.LogError($"Error happend while trying to delete indexes in Elasticsearch.");
+                    return false;
+                }
+            }
+            else
+            {
+                var workshops = await databaseService.GetByIds(ids).ConfigureAwait(false);
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs:146]: so we're getting workshop entities, convert them to dtos and then convert them to ES models. We are not using the DTOs in any way except to convert to ES. maybe a dedicated mapper Entity -> ES model will be a better solution?

**Review by DmyMi — APPROVED** [review]

**Review by provicevko — APPROVED** [review]
